packages/@react-spectrum/textfield/src/TextArea.tsx

```diff
@@ -26,31 +27,95 @@ function TextArea(props: SpectrumTextFieldProps, ref: RefObject<TextFieldRef>) {
 
   let inputRef = useRef<HTMLTextAreaElement>();
 
+  // Store styleProps to apply styles appropriately on height change.
+  let {styleProps} = useStyleProps(props);
+  let initialHeightsRef = useRef(null);
+
   let onHeightChange = useCallback(() => {
+    let input = inputRef.current;
+    let inputContainer = input.parentElement;
+    let field = ref.current.UNSAFE_getDOMNode();
+
+    // current offsetHeight of the textarea
+    let inputOffsetHeight = input.offsetHeight;
+
+    // Label or helpText height might change, so we have to recalculate
+    let labelHeight = labelPosition !== 'side' && !field.firstElementChild.contains(input) ? (field.firstElementChild as HTMLElement).offsetHeight : 0;
+    
+    // With labelPosition === 'side', the TextArea and HelpText will be contained within a wrapper.
+    let fieldWrapper = labelPosition === 'side' ? field.lastElementChild : field;
+    let helpTextHeight = !fieldWrapper.lastElementChild.contains(input) ? (fieldWrapper.lastElementChild as HTMLElement).offsetHeight : 0;
+
+    // store a ref containing the initial size values for comparison
+    if (initialHeightsRef.current === null) {
+      initialHeightsRef.current = {
+        inputHeight: inputOffsetHeight,
+        fieldHeight: field.offsetHeight,
+        diff: field.offsetHeight - inputOffsetHeight
+      };
+    }
+    let {inputHeight, fieldHeight, diff} = initialHeightsRef.current;
+    let calculatedFieldHeight = Math.max(inputHeight, inputOffsetHeight) + labelHeight + helpTextHeight;
     if (isQuiet) {
-      let input = inputRef.current;
       let prevAlignment = input.style.alignSelf;
       input.style.alignSelf = 'start';
       input.style.height = 'auto';
-      input.style.height = `${input.scrollHeight}px`;
+      input.style.height = `${Math.max(input.scrollHeight, inputHeight)}px`;
       input.style.alignSelf = prevAlignment;
+      inputContainer.style.flexGrow = '0';
+      for (const [key, value] of Object.entries(styleProps.style)) {
+        switch (key) {
+          case 'height':
+          case 'maxHeight':
+            field.style[key] = calculatedFieldHeight <= fieldHeight ? value : '';
+            break;
+        }
+      }
+    } else {  
+      for (const [key, value] of Object.entries(styleProps.style)) {
+        switch (key) {
+          case 'height':
+            field.style[key] = calculatedFieldHeight <= fieldHeight ? value : `${calculatedFieldHeight}px`;
+            
+            // When the textarea resizes smaller than the initial height, 
+            // its container should shrink, so that any helpText will align 
+            // to the bottom of the textarea.
+            inputContainer.style.flexGrow = inputOffsetHeight < inputHeight ? '0' : '';
+            break;
+          case 'minHeight':
+          case 'maxHeight':
+            input.style[key] = `calc(${value} - ${diff + (key === 'minHeight' ? -1 : 0) * helpTextHeight}px)`;
```

Comment:
Should this be `input.style[key] = `calc(${value} -  ${labelHeight}px - ${helpTextHeight}px)`;`? When I changed the `custom height size-2000,  labelPosition: side` HelpText story to be a TextArea w/ a `maxHeight="400px"` instead of a set height, I noticed that I couldn't expand the TextArea to the full 400px max height when I had the error help text displayed.

Reply:
I think you may be right about that.